

# Exploring and pre-processing a dataset using Pandas 






## Introduction

The goal is to demonstrate proficiency in utilizing the Pandas library to preprocess and analyze datasets effectively. This includes cleaning, organizing, and manipulating data stored in a Pandas DataFrame to prepare it for advanced data visualization and analysis. My capability to apply these skills highlights my ability to streamline workflows, extract meaningful insights, and prepare data for complex visual or statistical evaluations.


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Exploring Datasets with *pandas*](#0)<br>
2. [The Dataset: Immigration to Canada from 1980 to 2013](#1)<br>
3. [*pandas* Basics](#2) <br>
4. [*pandas* Intermediate: Indexing and Selection](#3) <br>
5. [*pandas* Filtering based on a criteria](#4)<br>
6. [*pandas* Sorting Values](#5)

</div>


# Exploring Datasets with *pandas* <a id="0"></a>

*pandas* is an essential data analysis toolkit for Python. From their [website](http://pandas.pydata.org/):
>*pandas* is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, **real world** data analysis in Python.

The *pandas* API Reference: http://pandas.pydata.org/pandas-docs/stable/api.html.


## The Dataset: Immigration to Canada from 1980 to 2013 <a id="1"></a>


Dataset Source: [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/development/desa/pd/data/international-migration-flows).

The dataset contains annual data on the flows of international immigrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. The current version presents data pertaining to 45 countries.

In this notebook, we will focus on the Canadian immigration data.

![Data Preview](canada_dataset_screenshot.png)

 The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">here</a>.

---


## *pandas* Basics<a id="2"></a>


The first thing we'll do is install **openpyxl** (formerly **xlrd**), a module that *pandas* requires to read Excel files.


In [54]:
!pip install openpyxl==3.1.0
# !pip install --upgrade openpyxl
# !pip show openpyxl

Next, we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [22]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library

Let's download and import our primary Canadian Immigration dataset using *pandas*'s `read_excel()` method.


In [23]:
import requests
import pandas as pd

# Download the xlsx file
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx'
response = requests.get(url)

if response.status_code == 200:
    with open('Canada.xlsx', 'wb') as f:
        f.write(response.content)
    print('File downloaded successfully!')
else:
    print('Failed to download file:', response.status_code)


File downloaded successfully!


In [24]:
import pandas as pd

try:
    df_can = pd.read_excel(
        'Canada.xlsx',
        sheet_name='Canada by Citizenship',
        skiprows=range(20),
        skipfooter=2,
        engine='openpyxl'
    )
    print('Data read into a pandas dataframe!')
except FileNotFoundError:
    print("File not found. Please check the file path.")
except ValueError as ve:
    print(f"Value error occurred: {ve}")
except Exception as e:
    print(f"An unexpected error occurred: {e}")
    

Data read into a pandas dataframe!


In [25]:
df_can.head(3)


Type    Coverage       OdName  AREA AreaName   REG          RegName  \
0  Immigrants  Foreigners  Afghanistan   935     Asia  5501    Southern Asia   
1  Immigrants  Foreigners      Albania   908   Europe   925  Southern Europe   
2  Immigrants  Foreigners      Algeria   903   Africa   912  Northern Africa   

   DEV             DevName  1980  ...  2004  2005  2006  2007  2008  2009  \
0  902  Developing regions    16  ...  2978  3436  3009  2652  2111  1746   
1  901   Developed regions     1  ...  1450  1223   856   702   560   716   
2  902  Developing regions    80  ...  3616  3626  4807  3623  4005  5393   

   2010  2011  2012  2013  
0  1758  2203  2635  2004  
1   561   539   620   603  
2  4752  4325  3774  4331  

[3 rows x 43 columns]

In [26]:
# view the bottom 5 rows
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

In [27]:
# getting basic information about the dataframe 
df_can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 43 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      195 non-null    object
 1   Coverage  195 non-null    object
 2   OdName    195 non-null    object
 3   AREA      195 non-null    int64 
 4   AreaName  195 non-null    object
 5   REG       195 non-null    int64 
 6   RegName   195 non-null    object
 7   DEV       195 non-null    int64 
 8   DevName   195 non-null    object
 9   1980      195 non-null    int64 
 10  1981      195 non-null    int64 
 11  1982      195 non-null    int64 
 12  1983      195 non-null    int64 
 13  1984      195 non-null    int64 
 14  1985      195 non-null    int64 
 15  1986      195 non-null    int64 
 16  1987      195 non-null    int64 
 17  1988      195 non-null    int64 
 18  1989      195 non-null    int64 
 19  1990      195 non-null    int64 
 20  1991      195 non-null    int64 
 21  1992      195 no

In [28]:
# To get the list of column headers, call upon the data frame's `columns` instance variable.
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

In [29]:
# to get the list of indices, use the `.index` instance variables.
df_can.index

RangeIndex(start=0, stop=195, step=1)

In [30]:
# columns and index are not in list formats
print(type(df_can.columns))
print(type(df_can.index))


<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


In [31]:
#To get the index and columns as lists,  use the `tolist()` method.
df_can.columns.tolist()


['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [32]:
df_can.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [33]:
print(type(df_can.columns.tolist()))
print(type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


In [34]:
# size of dataframe (rows, columns)
df_can.shape

(195, 43)

**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [35]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

Let's rename the columns so that they make sense. We can use `rename()` method by passing in a dictionary of old and new names as follows:


In [36]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 38 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Country    195 non-null    object
 1   Continent  195 non-null    object
 2   Region     195 non-null    object
 3   DevName    195 non-null    object
 4   1980       195 non-null    int64 
 5   1981       195 non-null    int64 
 6   1982       195 non-null    int64 
 7   1983       195 non-null    int64 
 8   1984       195 non-null    int64 
 9   1985       195 non-null    int64 
 10  1986       195 non-null    int64 
 11  1987       195 non-null    int64 
 12  1988       195 non-null    int64 
 13  1989       195 non-null    int64 
 14  1990       195 non-null    int64 
 15  1991       195 non-null    int64 
 16  1992       195 non-null    int64 
 17  1993       195 non-null    int64 
 18  1994       195 non-null    int64 
 19  1995       195 non-null    int64 
 20  1996       195 non-null    int64

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [41]:
df_can['Total'] = df_can.select_dtypes(include=['number']).sum(axis=1)
df_can.head()

Country Continent           Region             DevName  1980  1981  \
0     Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1         Albania    Europe  Southern Europe   Developed regions     1     0   
2         Algeria    Africa  Northern Africa  Developing regions    80    67   
3  American Samoa   Oceania        Polynesia  Developing regions     0     1   
4         Andorra    Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   
3     0     0     0     0  ...     0     1     0     0     0     0     0   
4     0     0     0     0  ...     0     1     1     0     0     0     0   

   2012  2013   Total  
0  2635  2004  175917  
1   620   603   47097  
2  3774  4331  208317  
3     0     0      18  
4     1     1      45  

[5 rows x 39 columns]

We can check to see how many null objects we have in the dataset as follows:


In [42]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


In [43]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

              Total  
count  1.950000e+02  
mean   9.860235e+04  
std    2.753565e+05  
min    3.000000e+00  
25%    2.856000e+03  
50%    1.505400e+04  
75%    6.671850e+04  
max    2.075712e+06  

[8 rows x 35 columns]

---
## *pandas* Intermediate: Indexing and Selection (slicing)<a id="3"></a>


### Select Column
**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.
```python
    df.column_name               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python 
    df[['column 1', 'column 2']]  # returns dataframe
```
---


Example: Let's try filtering on the list of countries ('Country').


In [44]:
df_can.Country  # returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [45]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers. 
# for the sake of consistency, we will convert all column names to string later on.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [46]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [49]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013   Total  
Afghanistan  2203  2635  2004  175917  
Albania       539   620   603   47097  
Algeria      4325  3774  4331  208317  

[3 rows x 38 columns]

In [50]:
# optional: to remove the name of the index
df_can.index.name = None

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:
    1. The full row data (all columns)
    2. For year 2013
    3. For years 1980 to 1985


In [51]:
# 1. the full row data (all columns)
df_can.loc['Japan']

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [52]:
# alternate methods
df_can.iloc[87]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [53]:
df_can[df_can.index == 'Japan']

Continent        Region            DevName  1980  1981  1982  1983  \
Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

       1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  2011  2012  \
Japan   246   198   248  ...  1067  1212  1250  1284  1194  1168  1265  1214   

       2013  Total  
Japan   982  83121  

[1 rows x 38 columns]

In [ ]:
# 2. for year 2013
df_can.loc['Japan', 2013]

In [ ]:
# alternate method
# year 2013 is the last column, with a positional index of 36
df_can.iloc[87, 36]

In [ ]:
# 3. for years 1980 to 1985
df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]]

In [ ]:
# Alternative Method
df_can.iloc[87, [3, 4, 5, 6, 7, 8]]

**Exercise:** Let's view the number of immigrants from **Haiti** for the following scenarios: <br>1. The full row data (all columns) <br>2. For year 2000 <br>3. For years 1990 to 1995


In [ ]:
df_can.loc['Haiti']
df_can.loc['Haiti', 2000]
df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]

<details><summary>Click here for a sample python solution</summary>

```python
   # 1. the full row data (all columns)
    df_can.loc['Haiti']
    #or
    df_can[df_can.index == 'Haiti']

    # 2. for year 2000
    df_can.loc['Haiti', 2000]
           
    #  3. for years 1990 to 1995
    df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]
 ```

</details>


Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index. 

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [ ]:
df_can.columns = list(map(str, df_can.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [ ]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

**Exercise:** Create a list named 'year' using map function for years ranging from 1990 to 2013. <br>Then extract the data series from the dataframe df_can for Haiti using year list. 


<details><summary>Click here for a sample python solution</summary>

```python
    #The correct answer is:
    year = list(map(str, range(1990, 2014)))
    haiti = df_can.loc['Haiti', year] # passing in years 1990 - 2013
    
```
</details>


### Filtering based on a criteria <a id="4"></a>
To filter the dataframe based on a condition, we simply pass the condition as a boolean vector. 

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [ ]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

In [ ]:
# 2. pass this condition into the dataFrame
df_can[condition]

In [ ]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

**Exercise:** Fetch the data where AreaName is 'Africa' and RegName is 'Southern Africa'. <br>Display the dataframe and find out how many instances are there?


<details><summary>Click here for a sample python solution</summary>

```python
    df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]
```
</details>


### Sorting Values of a Dataframe or Series <a id="5"></a><br>
You can use the `sort_values()` function is used to sort a DataFrame or a Series based on one or more columns. <br>You to specify the column(s) by which you want to sort and the order (ascending or descending). Below is the syntax to use it:-<br><br>
```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br><br>
col_nam - the column(s) to sort by. <br>
axis - axis along which to sort. 0 for sorting by rows (default) and 1 for sorting by columns.<br>
ascending - to sort in ascending order (True, default) or descending order (False).<br>
inplace - to perform the sorting operation in-place (True) or return a sorted copy (False, default).<br>
ignore_index - to reset the index after sorting (True) or keep the original index values (False, default).<br>


Let's sort out dataframe df_can on 'Total' column, in descending order to find out the top 5 countries that contributed the most to immigration to Canada. 


In [ ]:
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
top_5 = df_can.head(5)
top_5

**Exercise:** Find out top 3 countries that contributes the most to immigration to Canda in the year 2010. <br> Display the country names with the immigrant count in this year


<details><summary>Click here for a sample python solution</summary>

```python
    df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
    top3_2010 = df_can['2010'].head(3)
    top3_2010
```
</details>
